### PR TITLE
cache decoded line's size rather than recomputing at each stage

### PR DESCRIPTION
**Pull request by kevinushey**
This PR fixes an issue where attempting to search through files with long lines (e.g. base64 encoded content) would cause a quadratic re-computation of the number of UTF-8 characters in the line as matches were discovered in the line.

Rather than computing the number of UTF-8 characters from start to end on each match, we cache and update the size as we go along.

### CODE REVIEW

File: src/cpp/session/modules/SessionFind.cpp [hunk 1]
```diff
@@ -270,29 +270,32 @@ class GrepOperation : public boost::enable_shared_from_this<GrepOperation>
...
+         decodedLine.append(decoded);
          inputPos += match.position() + match.length();
-
-         size_t charSize;
-         Error error = string_utils::utf8Distance(decodedLine.begin(),
-                                                  decodedLine.end(),
+         
+         // count the number of UTF-8 characters processed
+         std::size_t charSize;
+         Error error = string_utils::utf8Distance(decoded.begin(),
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionFind.cpp]: One potential issue -- if the data is streamed in byte by byte, is it possible that we get 'half' of a UTF-8 character upon processing?

If so, we might want to keep the logic here as-is but limit the size we search.

**Comment by kevinushey** [comment]
@jjallaire what are your thoughts on this PR? If you'd prefer we don't touch the inner logic we can also just bail out after counting a certain amount of characters / bytes on a line.

**Comment by jjallaire** [comment]
Let's hold this one until after the preview next week. I'd definitely like
to come up with a solution that doesn't discard longer lines as I think
it's important that searches have integrity (as they are often lightweight
refactors and if we silently leave some results on the floor they can't be
trusted the way they should be).

**Comment by kevinushey** [comment]
@jjallaire let me know if you want to take a look at this one. This PR basically attempts to avoid some unintentional `O(N^2)` behavior that occurs as we re-scan the whole string on each iteration through.

**Comment by jjallaire** [comment]
This looks fine to me. Just to confirm: this approach *does not* discard longer lines at all? If so then feel free to merge (along with NEWS item)

**Comment by kevinushey** [comment]
Right, we do not discard longer lines -- we don't change any of the existing behavior; just the underlying implementation (we avoid re-scanning long lines on multiple matches)

**Pull request merged**